### ISSUE TITLE
Apache Directory StartTLS Configuration (Server Down. Result: 81)

### ISSUE THREAD

**Issue by tc-nathan** [flamencist/ldap4net#114]
Environment Details
 - OS: Windows 10 (client) Ubuntu 20.04 (server)
 - Ldap4net version: 2.7.11
 - .NET core version: 3.1
 - LDAP server: Apache Directory (https://directory.apache.org/)

Hi there, I'm getting the following error when attempting to enable StartTLS having just successfully tested without... 

    _System.Exception: Server Down. Server Down. Result: 81. Method: TransportLayerSecurityRequestHandler 
                    at LdapForNet.Native.LdapNative.ThrowIfError(SafeHandle ld, Int32 res, String method, IDictionary`2 details)
                    at LdapForNet.LdapConnection.SendRequest(DirectoryRequest directoryRequest, Int32& messageId)
                    at LdapForNet.LdapConnection.StartTransportLayerSecurity(Boolean trustAll)_

So if I comment out:
```
                    conn.SetClientCertificate(cert);
                    conn.StartTransportLayerSecurity(true);
```
it will work again but unencrypted...

This error will occur when cn.Bind is ran (see code below)

```
                var auth = Native.LdapAuthMechanism.SIMPLE;
                var host = "login.domain.tld";
                var @base = "OU=users,DC=login,DC=domain,DC=tld";
                var filter = "(objectclass=*)";   
                var port = 10389;

                string certPath = "C:\\path\\to\\where\\CAcertis\\login.domain.tld.cer";
                X509Certificate2 cert = new X509Certificate2(certPath);

                using (var conn = new LdapConnection())
                {
                    conn.Connect(host, port);
                    conn.SetClientCertificate(cert);
                    conn.StartTransportLayerSecurity(true);

                    var bindDN = "uid=admin,ou=system";
                    var bindPW = "password";

                    conn.Bind(auth, bindDN, bindPW);

                    IList<LdapEntry> entries;

                    entries = await conn.SearchAsync(@base, filter);
                    foreach (var ldapEntry in entries)
                    {
                        Console.WriteLine(ldapEntry);
                    }
                } 
```

As I'm testing on Windows I have installed the self-signed CA certificate locally which I extracted from a .ks (keystore) that was created on the directory server as part of following the official documentation. (https://directory.apache.org/apacheds/basic-ug/3.3-enabling-ssl.html) 

**note** its old documentation for LDAPS with incomplete documentation for StartTLS but I tweaked to work for connecting Apache Directory Studio (local) > Apache Directory (server) utilizing StartTLS successfully)
**note 2**: Apache Directory Studio is actually a client to administer from your local machine, with Apache Directory being the actual LDAP application.

In my Windows tests connecting my local dotnetcore application to the server, I attempted to: install the cert to the machine, user and also providing it with the        conn.SetClientCertificate(cert);       but I'm unfortunately always getting the above error.

I'd expect it to just work, but I'm not sure what I've done wrong and can't seem to figure it out. Most likely incorrect certificate configuration but my only working setup is where it's using Apache Directory Studio where you just give it the path to the .ks rather than extract and add the CA... 

Please help!! Hopefully someone has done this before, or has some helpful resource I can use :)

**Comment by DouglasKreider** [comment]
Same problem here. I know that TLS is set up properly on the server side, since it works with other clients. I'm using MMC on Windows to add the cert. I'm not completely sure I did that correctly. Is there a problem with the package, or is something else wrong?

**Comment by flamencist** [comment]
Did you try https://github.com/flamencist/ldap4net#connect-tls? Please attach any logs.

**Comment by DouglasKreider** [comment]
My file.

`using LdapForNet;
using ServerControl.Database.Models;
using ServerControl.Database.Services;
using System;
using System.Collections.Generic;
using System.Text;
using System.Security.Cryptography.X509Certificates;
using static LdapForNet.Native.Native;

namespace ServerControl.Email.Posfix_Services
{
    public class PostfixServerService
    {
        private readonly ApiConfiguration _credentials;

        public PostfixServerService(CredentialsService credentialsService)
        {
            _credentials = credentialsService.GetCredentials<ApiConfiguration>("ldap");
            Connection = new LdapConnection();
            Connection.Connect(new Uri(_credentials.Url));
            Connection.StartTransportLayerSecurity(true);
            Connection.Bind(LdapAuthMechanism.SIMPLE, "cn=Manager,dc=emypeople,dc=net", _credentials.AccessToken);
        }

        ~PostfixServerService()
        {
            Connection.Dispose();
        }

        public LdapConnection Connection { get; private set; }`

Server logs

May 17 11:33:22 dev-ldap.emypeople.net slapd[482]: conn=1790 fd=16 ACCEPT from IP=192.168.1.3:52823 (IP=0.0.0.0:389)
May 17 11:33:22 dev-ldap.emypeople.net slapd[482]: conn=1790 fd=16 closed (connection lost)
May 17 11:33:33 dev-ldap.emypeople.net slapd[482]: conn=1791 fd=16 ACCEPT from IP=192.168.1.3:52847 (IP=0.0.0.0:389)
May 17 11:33:33 dev-ldap.emypeople.net slapd[482]: conn=1791 fd=16 closed (connection lost)

As I study into it more, it seems likely it has to do with the TLS version or a wrongly installed cert.

**Comment by DouglasKreider** [comment]
The docs don't really say. Where is the cert supposed to be installed? I stepped through the code, and at least till the connection to the server, I can't find any mention of it getting the cert. Does that happen latter?

**Comment by DouglasKreider** [comment]
I believe I found the problem. I called _cn.StartTransportSecurity(false) and a Wire shark showed that the client was in the middle of configuring STARTTLS, and then called a reset, presumably because of the self signed cert. However, when I called _cn.StartTransportSecurity(true), Wire shark showed that the STARTTLS negotiation was never started. This seems to be happening because _cn.StartTransportSecurity() calls _cn.TrustAllCertificates() which in turn forces SSL to turn on. Our server is not configured to work with LDAPS, and so the connection was dropped. When I ran the code with the lines forcing SSL to be turned on commented out, I no longer got the Sever Down error.

What I would suggest is adding another method to the LdapNative classes only for trusting self signed STARTTLS certificates, not SSL ones. It would be identical to the method for allowing self signed SSL certificates, except it would not force SSL to be turned on.
